sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -137,6 +137,8 @@ object TypeCoercion {
     case (DateType, TimestampType) => Some(StringType)
     case (StringType, NullType) => Some(StringType)
     case (NullType, StringType) => Some(StringType)
+    case (n: NumericType, s: StringType) => Some(DoubleType)
+    case (s: StringType, n: NumericType) => Some(DoubleType)
```

Comment:
We intentionally not to do it, see https://github.com/apache/spark/pull/15880 .

But your use case is valid, I think casting string to the decimal type of the other side is too strict, how about
```
case (_: StringType, d: DecimalType) => DecimalPrecision.widerDecimalType(DecimalType.SYSTEM_DEFAULT, d)
```

cc @gatorsmile

Reply:
This solved DecimalType but  may also have known error.
For example:
`select '1.01' > 1.0;` // false 

Is it possible to use DecimalType.SYSTEM_DEFAULT?

Reply:
How about we cast both sides to double for string and decimal combo? We can't find a proper decimal precision anyway.

Reply:
Because `select '1.1' >1` returns false, I prefer casting all NumericType to double like hive.
Therefore, casting decimal to double looks better for me.
But, in our cluster, many users write SQL like `select '1.1' > 1`, this compatibility brings great difficulties to transferring hive task to spark task. So, don't we really need to think about casting all NumericType to double?